util/formats/newfmt/types.go

```diff
@@ -0,0 +1,217 @@
+	Producer string `json:"producer"`
+	Version  string `json:"version"`
+	// The `Token` is not documented in the spec.  It is a makeshift authorization token (API token)
+	// to be used with Envelope.Data.Attributes.Cluster, it proves that the producer of the datum
+	// was authorized to produce data for that cluster name.
+	Token string `json:"token"`
+}
+
+type ErrorObject struct {
+	Time   string `json:"time"`
+	Detail string `json:"detail"`
+}
+
+type KVPair struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+// JSON representation can be read with ConsumeJSONSysinfo().
+
+type SysinfoEnvelope struct {
+	Data   *SysinfoData   `json:"data"`
+	Errors []ErrorObject  `json:"errors"`
+	Meta   MetadataObject `json:"meta"`
+}
+
+type SysinfoData struct {
+	Type       string            `json:"type"`
+	Attributes SysinfoAttributes `json:"attributes"`
+}
+
+type SysinfoAttributes struct {
+	Time        string                   `json:"time"`
+	Cluster     string                   `json:"cluster"`
+	Node        string                   `json:"node"`
+	OsName      string                   `json:"os_name"`
+	OsRelease   string                   `json:"os_release"`
+	Cores       []SysinfoCoreModel       `json:"cores"`
+	Memory      uint64                   `json:"memory"`
+	Description string                   `json:"description"`
+	TopoSVG     string                   `json:"topo_svg"` // Base64-encodes SVG
+	Cards       []SysinfoGpuCard         `json:"cards"`
+	Software    []SysinfoSoftwareVersion `json:"software"`
+}
+
+type SysinfoCoreModel struct {
+	Index    uint64 `json:"index"`
+	Physical uint64 `json:"physical"`
+	Model    string `json:"model"`
+}
+
+type SysinfoGpuCard struct {
+	Index         uint64 `json:"index"`
+	UUID          string `json:"uuid"`
+	Address       string `json:"address"`
+	Manufacturer  string `json:"manufacturer"`
+	Model         string `json:"model"`
+	Architecture  string `json:"architecture"`
+	Driver        string `json:"driver"`
+	Firmware      string `json:"firmware"`
+	Memory        uint64 `json:"memory"`
+	PowerLimit    uint64 `json:"power_limit"`
+	MaxPowerLimit uint64 `json:"max_power_limit"`
+	MinPowerLimit uint64 `json:"min_power_limit"`
+	MaxCEClock    uint64 `json:"max_ce_clock"`
+	MaxMemClock   uint64 `json:"max_memory_clock"`
+}
+
+type SysinfoSoftwareVersion struct {
+	Key     string `json:"key"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
+
+// JSON representation can be read with ConsumeJSONSamples().
+
+type SampleEnvelope struct {
+	Data   *SampleData    `json:"data"`
+	Errors []ErrorObject  `json:"errors"`
+	Meta   MetadataObject `json:"meta"`
+}
+
+type SampleData struct {
+	Type       string           `json:"type"`
+	Attributes SampleAttributes `json:"attributes"`
+}
+
+type SampleAttributes struct {
+	Time    string        `json:"time"`
+	Cluster string        `json:"cluster"`
+	Node    string        `json:"node"`
+	Attrs   []KVPair      `json:"attrs"`
```

Comment:
I don't like "attrs" both because it is an abbreviation and because it is confusing vis-a-vis "attributes", of which it is part.  Could this be "properties"?  Do we even want it here?  In slurm-monitor this carries eg the Python version.  Is that a relevant attribute of the sample data, as here, or is it metadata?  It depends.  If we want to encode the command line switches, which affect the data, it's definitely relevant here.  Is it maybe "environment"? "invocation"?

Reply:
In the spec this has moved to metadata, which makes some type of sense since eg the producer name and version are there, and the data envisioned for "attrs" is that type of data.